### ISSUE TITLE
ALOps steps and Antivirus

### ISSUE THREAD

**Issue by gintuoklis** [HodorNV/ALOps#108]
Hey Guys, 

We are experiencing these kind of errors with ALOps steps on one of the environment with antivirus software installed.

Standard AzureDevops steps are working fine (Download Artifacts, Checkout, Download package)

This environment has two antiviruses running (don't ask why)
SOPHOS Endpoint Security and Control, version 10.8
and 
Windows Defender 

It seems that Defender blocks it:
![image](https://user-images.githubusercontent.com/16342128/78902382-02cfe800-7a7a-11ea-934b-a492b3a4721e.png)

```
2020-04-09T11:26:30.0909960Z ##[section]Starting: Dev License Import
2020-04-09T11:26:30.1039454Z ==============================================================================
2020-04-09T11:26:30.1039812Z Task         : ALOps License Import
2020-04-09T11:26:30.1040055Z Description  : ALOps License Import
2020-04-09T11:26:30.1040277Z Version      : 1.428.1196
2020-04-09T11:26:30.1040495Z Author       : Hodor
2020-04-09T11:26:30.1040738Z Help         : Import Business Central license (.flf).
2020-04-09T11:26:30.1041314Z ==============================================================================
2020-04-09T11:26:33.8303223Z ##[error]At C:\agent\_work\_tasks\ALOpsLicenseImport_1da752a0-03bf-429c-a239-b5f06d43d629\1.428.1196\dcb2f576-3e99-440f-b336-5e161244c878.ps1:1 char:1
+ . ".\ps_modules\alops\_AlopsAmsiController.ps1"
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
This script contains malicious content and has been blocked by your antivirus software.
2020-04-09T11:26:38.6279153Z ##[section]Finishing: Dev License Import
```

```
2020-04-09T11:26:38.6498809Z ##[section]Starting: ALOps Import FOB
2020-04-09T11:26:38.6630581Z ==============================================================================
2020-04-09T11:26:38.6630973Z Task         : ALOps Import FOB
2020-04-09T11:26:38.6631199Z Description  : ALOps Import FOB
2020-04-09T11:26:38.6631417Z Version      : 1.428.1196
2020-04-09T11:26:38.6631665Z Author       : Hodor
2020-04-09T11:26:38.6631899Z Help         : Import objects from .FOB file.
2020-04-09T11:26:38.6632241Z ==============================================================================
2020-04-09T11:26:40.6128937Z ##[error]At line:1 char:1
+  ([RuNTImE.INTEroPsErVICeS.MArSHAL]::pTRtOsTRIngbSTr( [RUNTIMe.intERO ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
This script contains malicious content and has been blocked by your antivirus software.
2020-04-09T11:26:42.2704256Z ##[section]Finishing: ALOps Import FOB


```

**Comment by gintuoklis** [comment]
Additional question:
Let's we would like to use deployment pipelines on customer environment, since we have no access to your code, how can we trust your tasks that they do only the thing it claims? :) Especially, when self-hosted agents are already running in elevated mode (otherwise no NAV Admin shell tasks can be executed)

**Comment by AdminHodor** [comment]
Dear @gintuoklis ,

>>It seems that Defender blocks it:
Please try setting the Environment variable [alops_disable_amsi] = true.
This will disable the windows AMSI check for the Build/Release

>>This environment has two antiviruses running (don't ask why)
Please know, for build machines it is highly recommended to NOT run any antivirus or Defender at all. 
To secure a build agent, it's best to isolate it in a Sandbox environment or DMZ.

Concerning Release flows, isolation of the agent is nearly impossible, but most risk can be mitigated in the way the entire system is setup. Especially when using docker, these issues are quite easy to mitigate.

>>since we have no access to your code, how can we trust your tasks that they do only the thing it claims?
As is for any proprietary software....

>> deployment pipelines on customer environment
We are aware this is currently the only way for including customer deployments in the pipeline, but do know this is very bad practice. You have a Full-Admin agent running at the customers, connected to your devops. The customer this way exposes himself to the same risks you ask us about.

But,.... we do have a solution in the Pipeline that will enable you to run Customer-Release pipelines without having a DevOps agent present at the customer. (API Deploy). The ALOps task for this is already done, we are finalizing the "BC Service" that is required to have the API-Deploy available on a onprem Business Central installation (feature is not available onprem).

In short:
  - Please try setting boolean Environment variable [**alops_disable_amsi**] = true
  - Next Major release (v2.0) will be fully dotnet instead of Powershell, which will bypass the false virusscanner claims
  - API deploy is almost ready, no DevOps agent at the customer will be required if the have a public API port to Business Central.

Kind regards,

Bart Bourgeois

**Comment by gintuoklis** [comment]
Thanks, for you answers @AdminHodor .
At the moment we are still dealing with mixed mode (fobs, and apps), therefore agents are necessary :( 

Can you give more details of how and where i should set "alops_disable_amsi"?

**Comment by AdminHodor** [comment]
Hello @gintuoklis ,

It's a regular pipeline variable, nothing special.

In Yaml you can set it like this:
```
variables:
  alops_disable_amsi: true
```
Or in the Build/Release pipeline you can use the Variable dialog.
![image](https://user-images.githubusercontent.com/51383893/78979366-03b95600-7b1c-11ea-8589-fc5a15261d1b.png)

Kind regards,

**Comment by gintuoklis** [comment]
Can you tell, what exactly this flag will do. Will it add exception for your scripts? Or completely disable Defender?

**Comment by AdminHodor** [comment]
The flag disables AMSI (Anti Malware Scan Interface) for the current Task per Session, it's not a global rewrite of configuration change. If you have an agent where you don't want to disable this, and you are concerned about the admin permissions, you really should think twice about installing a DevOps Agent anyways.

We definitely will keep you posted on the API-Deploy, which should the mitigate the need for DevOps agent on customers servers, or critical servers. 

Kind regards,
